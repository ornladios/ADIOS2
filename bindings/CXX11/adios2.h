/*
 * Distributed under the OSI-approved Apache License, Version 2.0.  See
 * accompanying file Copyright.txt for details.
 *
 * adios2.h : single unique header for applications using ADIOS2 C++11 bindings
 *
 *  Created on: Jun 4, 2018
 *      Author: William F Godoy godoywf@ornl.gov
 */

#ifndef ADIOS2_BINDINGS_CXX11_H_
#define ADIOS2_BINDINGS_CXX11_H_

#include "adios2/ADIOSConfig.h"
#include "adios2/ADIOSMacros.h"
#include "adios2/ADIOSTypes.h"

#include "cxx11/ADIOS.h"
#include "cxx11/Attribute.h"
#include "cxx11/Engine.h"
#include "cxx11/IO.h"
#include "cxx11/Operator.h"
#include "cxx11/Types.h"
#include "cxx11/Variable.h"
#include "cxx11/fstream/ADIOS2fstream.h"

#endif /* ADIOS2_BINDINGS_CXX11_H_ */
