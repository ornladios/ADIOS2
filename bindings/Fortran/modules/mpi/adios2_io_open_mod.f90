!
! Distributed under the OSI-approved Apache License, Version 2.0.  See
!  accompanying file Copyright.txt for details.
!
!  adios2_io_open_mod.f90 : ADIOS2 Fortran bindings for IO class open function
!
!   Created on: Mar 13, 2017
!       Author: William F Godoy godoywf@ornl.gov
!

module adios2_io_open_mod
    use adios2_parameters_mod
    implicit none

    interface adios2_open
        module procedure adios2_open_old_comm
        module procedure adios2_open_new_comm
    end interface

contains

    subroutine adios2_open_old_comm(engine, io, name, adios2_mode, ierr)
        type(adios2_engine), intent(out) :: engine
        type(adios2_io), intent(in) :: io
        character*(*), intent(in) :: name
        integer, intent(in) :: adios2_mode
        integer, intent(out) :: ierr

        engine%mode = adios2_mode
        call adios2_open_f2c(engine%f2c, io%f2c, TRIM(ADJUSTL(name))//char(0), &
                             adios2_mode, ierr)

        if( ierr == 0 ) then
            engine%valid = .true.
            engine%name = name
            call adios2_engine_get_type_f2c(engine%type, engine%f2c, ierr)
            engine%mode = adios2_mode
        end if

    end subroutine

    subroutine adios2_open_new_comm(engine, io, name, adios2_mode, comm, ierr)
        type(adios2_engine), intent(out) :: engine
        type(adios2_io), intent(in) :: io
        character*(*), intent(in) :: name
        integer, intent(in) :: adios2_mode
        integer, intent(in) :: comm
        integer, intent(out) :: ierr

        call adios2_open_new_comm_f2c(engine%f2c, io%f2c, &
                                      TRIM(ADJUSTL(name))//char(0), &
                                      adios2_mode, comm, ierr)

        if( ierr == 0 ) then
            engine%valid = .true.
            engine%name = name
            call adios2_engine_get_type_f2c(engine%type, engine%f2c, ierr)
            engine%mode = adios2_mode
        end if

    end subroutine

end module
