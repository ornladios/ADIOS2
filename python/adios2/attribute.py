"""License:
  Distributed under the OSI-approved Apache License, Version 2.0.  See
  accompanying file Copyright.txt for details.
"""


class Attribute:
    """High level representation of the Attribute class in the adios2.bindings"""

    def __init__(self, io, name, *args, **kwargs):
        self.impl = io.DefineAttribute(name, *args, **kwargs)

    @property
    def impl(self):
        """Bindings implementation of the class"""
        return self._impl

    @impl.setter
    def impl(self, implementation):
        self._impl = implementation

    def __eq__(self, other):
        if isinstance(other, Attribute):
            return self.name() == other.name()
        return False

    def name(self):
        """
        Name of the Attribute

        Returns:
            Name of the Attribute as a str.
        """
        return self.impl.Name()

    def type(self):
        """
        Type of the Attribute

        Returns:
            Type of the Attribute as a str.
        """
        return self.impl.Type()

    def single_value(self):
        """
        True if the attribute is a single value, False if it is an array

        Returns:
            True or False.
        """
        return self.impl.SingleValue()

    def data(self):
        """
        Content of the Attribute

        Returns:
            Content of the Attribute as a non string.
        """
        if self.single_value():
            return self.impl.Data()[0]
        return self.impl.Data()

    def data_string(self):
        """
        Content of the Attribute

        Returns:
            Content of the Attribute as a str.
        """
        if self.single_value():
            return self.impl.DataString()[0]
        return self.impl.DataString()
