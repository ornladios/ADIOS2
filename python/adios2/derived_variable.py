"""License:
  Distributed under the OSI-approved Apache License, Version 2.0.  See
  accompanying file Copyright.txt for details.
"""


class DerivedVariable:
    """High level representation of the DerivedVariable class in the adios2.bindings"""

    def __init__(self, implementation):
        self.impl = implementation

    @property
    def impl(self):
        """Bindings implementation of the class"""
        return self._impl

    @impl.setter
    def impl(self, implementation):
        self._impl = implementation

    def __eq__(self, other):
        if isinstance(other, DerivedVariable):
            return self.name() == other.name()
        return False

    def type(self):
        """
        Type of the DerivedVariable

        Returns:
            str: Type of the DerivedVariable.
        """
        return self.impl.Type()

    def name(self):
        """
        Name of the DerivedVariable

        Returns:
            str: Name of the DerivedVariable.
        """
        return self.impl.Name()
