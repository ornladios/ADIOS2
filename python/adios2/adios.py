"""License:
  Distributed under the OSI-approved Apache License, Version 2.0.  See
  accompanying file Copyright.txt for details.
"""

from adios2 import bindings
from adios2.io import IO
from adios2.operator import Operator


class Adios:
    """High level representation of the ADIOS class in the adios2.bindings"""

    def __init__(self, config_file=None, comm=None):
        if comm and not bindings.is_built_with_mpi:
            raise RuntimeError("Cannot use MPI since ADIOS2 was built without MPI support")

        if config_file:
            if comm:
                self.impl = bindings.ADIOS(config_file, comm)
            else:
                self.impl = bindings.ADIOS(config_file)
        else:
            if comm:
                self.impl = bindings.ADIOS(comm)
            else:
                self.impl = bindings.ADIOS()

    @property
    def impl(self):
        """Bindings implementation of the class"""
        return self._impl

    @impl.setter
    def impl(self, implementation):
        self._impl = implementation

    def declare_io(self, name):
        """
        Declare IO instance

        Args:
            name (str): IO instance name
        """
        return IO(self.impl.DeclareIO(name), name, self)

    def at_io(self, name):
        """
        Inquire IO instance

        Args:
            name (str): IO instance name
        Returns:
            IO: IO instance
        """
        io = None
        io_instance = self.impl.AtIO(name)
        if io_instance:
            io = IO(io_instance, name, self)
        return io

    def remove_io(self, name):
        """
        Remove IO instance

        Args:
            name (str): IO instance name
        """
        self.impl.RemoveIO(name)

    def remove_all_ios(self):
        """Remove all IO instance"""
        self.impl.RemoveAllIOs()

    def define_operator(self, name, kind, parameters={}):
        """
        Add an operation (operator).

        Args:
            name (str): name of the operator.
            kind (str): name type of the operation.
            params (dict): parameters as a form of a dict for the operation.
        """
        return Operator(self.impl.DefineOperator(name, kind, parameters), name)

    def inquire_operator(self, name):
        """
        Add an operation (operator).

        Args:
            name (str): name of the operator.
        Return:
            Operator: requested operator
        """
        operator = None
        operator_instance = self.impl.InquireOperator(name)
        if operator_instance:
            operator = Operator(self.impl.InquireOperator(name), name)

        return operator

    def flush_all(self):
        """
        Flush all IO instances
        """
        self.impl.FlushAll()
