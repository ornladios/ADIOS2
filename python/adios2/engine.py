"""License:
  Distributed under the OSI-approved Apache License, Version 2.0.  See
  accompanying file Copyright.txt for details.
"""

import numpy as np

from adios2 import bindings


class Engine:
    """High level representation of the Engine class in the adios2.bindings"""

    def __init__(self, implementation):
        self.impl = implementation

    @property
    def impl(self):
        """Bindings implementation of the class"""
        return self._impl

    @impl.setter
    def impl(self, implementation):
        self._impl = implementation

    def __enter__(self):
        return self

    def __exit__(self, *exc):
        self.close()

    def close(self, transport_index=-1):
        """
        Close an transports of the engine

        Args:
            transportIndex (int): if -1 close all transports
        """
        self.impl.Close(transport_index)

    def steps(self):
        """Returns number of available steps"""
        return self.impl.Steps()

    def current_step(self):
        """Returns the current step"""
        return self.impl.CurrentStep()

    def begin_step(self, *args, **kwargs):
        """Start step"""
        return self.impl.BeginStep(*args, **kwargs)

    def end_step(self):
        """End step"""
        self.impl.EndStep()

    def between_step_pairs(self):
        """End step"""
        return self.impl.BetweenStepPairs()

    def all_blocks_info(self, name):
        """
        Returns a list of BlocksInfo for a all steps

        Args:
            name (str): variable name

        Returns:
            list of BlockInfos
        """
        output = []
        for step in range(0, self.steps()):
            output.append(self.blocks_info(name, step))
        return output

    def blocks_info(self, name, step):
        """
        Returns a BlocksInfo for a given step

        Args:
            name (str): variable name
            step (int): step in question

        Returns:
            list of dicts describing each BlockInfo
        """
        return self.impl.BlocksInfo(name, step)

    def put(self, variable, content, mode=bindings.Mode.Deferred):
        """
        Puts content in a variable

        Parameters
            name
                variable name

            content
                variable data values

            mode
                when to perform the communication, (Deferred or Asynchronous).
        """
        if isinstance(content, np.ndarray):
            self.impl.Put(variable.impl, content, mode)
        elif isinstance(content, str):
            self.impl.Put(variable.impl, content)
        elif isinstance(content, list):
            self.impl.Put(variable.impl, content)
        elif not hasattr(content, "__len__"):
            content = np.array([content])
            self.impl.Put(variable.impl, content)
        else:
            raise ValueError

    def perform_puts(self):
        """Perform the puts calls"""
        self.impl.PerformPuts()

    def perform_data_write(self):
        """Perform the transport data writes"""
        self.impl.PerformDataWrite()

    def get(self, variable, content=None, mode=bindings.Mode.Sync):
        """
        Gets the content of a variable

        Parameters
            name
                variable name

            content
                output variable data values

            mode
                when to perform the communication, (Deferred or Asynchronous).
        Returns
            Content of the variable when the content argument is omitted.
        """
        if isinstance(content, np.ndarray):
            self.impl.Get(variable.impl, content, mode)
            return None

        return self.impl.Get(variable.impl, mode)

    def perform_gets(self):
        """Perform the gets calls"""
        self.impl.PerformGets()

    def lock_reader_selections(self):
        """Locks the data selection for read"""
        self.impl.LockReaderSelections()

    def lock_writer_definitions(self):
        """Locks the data selection for write"""
        self.impl.LockWriterDefinitions()

    def flush(self, transport_index=-1):
        """Flush all transports attached to this Engine instance"""
        self.impl.Flush(transport_index)
