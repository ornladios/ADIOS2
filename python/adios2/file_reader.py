"""License:
  Distributed under the OSI-approved Apache License, Version 2.0.  See
  accompanying file Copyright.txt for details.
"""

from functools import singledispatchmethod
from adios2 import Stream, IO


# pylint: disable=W0221
class FileReader(Stream):
    """High level implementation of the FileReader class for read Random access mode"""

    def __repr__(self):
        return f"<adios.file named {self._io_name}>"

    @singledispatchmethod
    def __init__(self, path, comm=None):
        super().__init__(path, "rra", comm)

    # e.g. FileReader(io: adios2.IO, path, mode)
    # pylint: disable=E1121
    @__init__.register(IO)
    def _(self, io: IO, path, comm=None):
        super().__init__(io, path, "rra", comm)

    # pylint: enable=E1121
