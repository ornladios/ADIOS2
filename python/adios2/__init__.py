"""The ADIOS2 high-level API module
License:
  Distributed under the OSI-approved Apache License, Version 2.0.  See
  accompanying file Copyright.txt for details.
"""

import adios2.bindings

from adios2.adios import *
from adios2.attribute import *
from adios2.engine import *
from adios2.io import *
from adios2.operator import *
from adios2.stream import *
from adios2.variable import *
from adios2.file_reader import *
from adios2.bindings import LocalValueDim

__license__ = "Apache-2.0"
__version__ = adios2.bindings.__version__
is_built_with_mpi = adios2.bindings.is_built_with_mpi
