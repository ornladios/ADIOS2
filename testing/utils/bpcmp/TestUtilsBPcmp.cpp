#include <algorithm>
#include <ios>
#include <iostream>
#include <numeric>
#include <stdexcept>
#include <vector>

#include <adios2.h>
#if ADIOS2_USE_MPI
#include <mpi.h>
#endif

int main(int argc, char *argv[])
{
    int rank, size;
#if ADIOS2_USE_MPI
    int provided;
    MPI_Init_thread(&argc, &argv, MPI_THREAD_MULTIPLE, &provided);
    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
    MPI_Comm_size(MPI_COMM_WORLD, &size);
#else
    rank = 0;
    size = 1;
#endif

    const size_t Nx = 2, Ny = 3;
    const size_t steps = 2;
    std::vector<float> commonArray(Nx * Ny);
    std::iota(commonArray.begin(), commonArray.end(), 4.0f);

#if ADIOS2_USE_MPI
    adios2::ADIOS adios(MPI_COMM_WORLD);
#else
    adios2::ADIOS adios;
#endif

    /* PRIMARY file */
    {
        adios2::IO bpIO = adios.DeclareIO("BPcmp1");

        auto bpFloats = bpIO.DefineVariable<float>("bpFloats", {size * Nx, Ny}, {rank * Nx, 0},
                                                   {Nx, Ny}, adios2::ConstantDims);

        auto bpStep = bpIO.DefineVariable<unsigned int>("bpStep");

        // Create the raw file with the 2D raw array and step information
        adios2::Engine bpWriter = bpIO.Open("TestBPcmp_set1.bp", adios2::Mode::Write);

        for (size_t i = 0; i < steps; i++)
        {
            bpWriter.BeginStep();
            bpWriter.Put(bpFloats, commonArray.data());
            bpWriter.Put(bpStep, static_cast<unsigned int>(i));
            bpWriter.EndStep();

            std::transform(commonArray.begin(), commonArray.end(), commonArray.begin(),
                           [](float &v) { return v + 1; });
        }

        bpWriter.Close();
    }

    /* SECONDARY file */
    {
        adios2::IO bpIO = adios.DeclareIO("BPcmp2");

        auto bpFloats = bpIO.DefineVariable<float>("bpFloats", {size * Nx, Ny}, {rank * Nx, 0},
                                                   {Nx, Ny}, adios2::ConstantDims);

        auto bpStep = bpIO.DefineVariable<int>("bpStep", {1}, {0}, {1});

        // Create the raw file with the 2D raw array and step information
        adios2::Engine bpWriter = bpIO.Open("TestBPcmp_set2.bp", adios2::Mode::Write);

        for (size_t i = 0; i < steps; i++)
        {
            bpWriter.BeginStep();
            bpWriter.Put(bpFloats, commonArray.data());
            bpWriter.Put(bpStep, static_cast<int>(i));
            bpWriter.EndStep();
        }
        bpWriter.BeginStep();
        bpWriter.Put(bpStep, static_cast<int>(steps));
        bpWriter.EndStep();

        bpWriter.Close();
    }

#if ADIOS2_USE_MPI
    MPI_Finalize();
#endif

    return 0;
}
