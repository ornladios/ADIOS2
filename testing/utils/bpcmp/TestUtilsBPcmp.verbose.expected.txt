COMPARE : FILE : TestBPcmp_set1.bp : FILE : TestBPcmp_set2.bp
DIFFER : NUM_STEPS : 2 <> 3 : TestBPcmp_set1.bp : TestBPcmp_set2.bp
DIFFER : VARIABLE : bpFloats : STEP : 0 : POSITION : (0, 0) : VALUE : 4.0 <> 6.0 : TestBPcmp_set1.bp : TestBPcmp_set2.bp
DIFFER : VARIABLE : bpFloats : STEP : 0 : POSITION : (0, 1) : VALUE : 5.0 <> 7.0 : TestBPcmp_set1.bp : TestBPcmp_set2.bp
DIFFER : VARIABLE : bpFloats : STEP : 0 : POSITION : (0, 2) : VALUE : 6.0 <> 8.0 : TestBPcmp_set1.bp : TestBPcmp_set2.bp
DIFFER : VARIABLE : bpFloats : STEP : 0 : POSITION : (1, 0) : VALUE : 7.0 <> 9.0 : TestBPcmp_set1.bp : TestBPcmp_set2.bp
DIFFER : VARIABLE : bpFloats : STEP : 0 : POSITION : (1, 1) : VALUE : 8.0 <> 10.0 : TestBPcmp_set1.bp : TestBPcmp_set2.bp
DIFFER : VARIABLE : bpFloats : STEP : 0 : POSITION : (1, 2) : VALUE : 9.0 <> 11.0 : TestBPcmp_set1.bp : TestBPcmp_set2.bp
DIFFER : VARIABLE : bpFloats : STEP : 1 : POSITION : (0, 0) : VALUE : 5.0 <> 6.0 : TestBPcmp_set1.bp : TestBPcmp_set2.bp
DIFFER : VARIABLE : bpFloats : STEP : 1 : POSITION : (0, 1) : VALUE : 6.0 <> 7.0 : TestBPcmp_set1.bp : TestBPcmp_set2.bp
DIFFER : VARIABLE : bpFloats : STEP : 1 : POSITION : (0, 2) : VALUE : 7.0 <> 8.0 : TestBPcmp_set1.bp : TestBPcmp_set2.bp
DIFFER : VARIABLE : bpFloats : STEP : 1 : POSITION : (1, 0) : VALUE : 8.0 <> 9.0 : TestBPcmp_set1.bp : TestBPcmp_set2.bp
DIFFER : VARIABLE : bpFloats : STEP : 1 : POSITION : (1, 1) : VALUE : 9.0 <> 10.0 : TestBPcmp_set1.bp : TestBPcmp_set2.bp
DIFFER : VARIABLE : bpFloats : STEP : 1 : POSITION : (1, 2) : VALUE : 10.0 <> 11.0 : TestBPcmp_set1.bp : TestBPcmp_set2.bp
DIFFER : VARIABLE : bpStep : NUM_STEPS : 2 <> 3 : TestBPcmp_set1.bp : TestBPcmp_set2.bp
DIFFER : VARIABLE : bpStep : TYPE : uint32_t <> int32_t : TestBPcmp_set1.bp : TestBPcmp_set2.bp
DIFFER : VARIABLE : bpStep : SINGLE_VALUE : true <> false : TestBPcmp_set1.bp : TestBPcmp_set2.bp
TOTAL_DIFFERENCES : 16 : FILE : TestBPcmp_set1.bp : FILE : TestBPcmp_set2.bp
